File: tests/oracle/funcdata_scopelocal_wrap_1204.rs
// Locked Ghidra 12.0.4 ScopeInternal::findOverlap wrap-domain oracle for
// FUNCDATA-SCOPELOCALOVERFLOW-0001 / FUNCDATA-SCOPELOCAL-WRAP-0001
// (Rust side).
//
// Mirrors tests/oracle/funcdata_scopelocal_wrap_1204.cc record-for-record:
// the same address-tied symbol table (top8 = [0xfffffffffffffff8,
// 0xffffffffffffffff], low8 = [0x100, 0x107]) queried through
// rugra::funcdata::scope_local_find_overlap — the LocalSymbol-granular
// projection ScopeInternal::findOverlap (database.cc:2392) that
// Funcdata::syncVarnodesWithSymbols (funcdata_varnode.cc:951) drives.
// The point of the fixture is the uint8 modular domain of
// `addr.getOffset()+size-1` (database.cc:2397): the pre-fix Rust trapped
// with `attempt to add with overflow` on the debug profile at these exact
// queries and missed the top-of-space record with its `p < first+size`
// containment form when first+size wrapped to 0.

use rugra::funcdata::scope_local_find_overlap;
use rugra::space::AddressSpace;
use rugra::varmap::{symbol_category, LocalSymbol, ScopeLocal};

fn add_addrtied_symbol(scope: &mut ScopeLocal, name: &str, start: u64, size: i32) {
    // Scope::addSymbol with an invalid usepoint: address-tied storage,
    // empty uselimit, subsort (0,0) (database.cc:1149-1150) — modeled by
    // usepoint == None on the LocalSymbol projection.
    let mut sym = LocalSymbol::new(name, start, size, None, symbol_category::NO_CATEGORY);
    sym.space = AddressSpace::Stack;
    sym.usepoint = None;
    scope.symbols.push(sym);
}

fn dump_overlap(case: &str, scope: &ScopeLocal, offset: u64, size: i32) {
    let entry = scope_local_find_overlap(scope, AddressSpace::Stack, offset, size);
    let line = match entry {
        None => format!("case={case}|kind=overlap|query={offset:#x}:{size}|up=inv|result=null"),
        Some(sym) => {
            let last = sym.start.wrapping_add(sym.size as u64).wrapping_sub(1);
            format!(
                "case={case}|kind=overlap|query={offset:#x}:{size}|up=inv|result={}|first={:#x}|last={:#x}|off=0|sz={}",
                sym.name, sym.start, last, sym.size
            )
        }
    };
    println!("{line}");
}

fn main() {
    let mut scope = ScopeLocal::new();
    scope.space = AddressSpace::Stack;
    add_addrtied_symbol(&mut scope, "top8", 0xffff_ffff_ffff_fff8, 8);
    add_addrtied_symbol(&mut scope, "low8", 0x100, 8);

    dump_overlap("wrap_top_full", &scope, 0xffff_ffff_ffff_fff8, 8);
    dump_overlap("wrap_top_last_byte", &scope, 0xffff_ffff_ffff_ffff, 1);
    dump_overlap("wrap_top_from_below", &scope, 0xffff_ffff_ffff_fff6, 4);
    dump_overlap("neg_size_below", &scope, 0x1000, -8);
    dump_overlap("zero_size_top", &scope, 0xffff_ffff_ffff_fff8, 0);
    dump_overlap("low_hit", &scope, 0x104, 4);
    dump_overlap("nonoverlap_above", &scope, 0x10, 8);
}

File: tests/oracle/funcdata_scopelocal_wrap_1204.cc
/*
 * Locked Ghidra 12.0.4 ScopeInternal::findOverlap wrap-domain oracle for
 * FUNCDATA-SCOPELOCALOVERFLOW-0001 / FUNCDATA-SCOPELOCAL-WRAP-0001.
 *
 * database.cc:2397 evaluates `addr.getOffset()+size-1` in the uint8
 * (uint64) modular domain: an int4 size converts by sign extension and
 * both operators wrap (C++ unsigned arithmetic). Stack-space offsets near
 * 2^64 — negative stack slots, e.g. a canary at stack -8 stored as
 * 0xfffffffffffffff8 — make the add wrap, and a size 0 query makes the
 * sub wrap one below the point. These cases pin the oracle's answers on
 * the wrap boundary, where a checked (non-modular) Rust translation traps
 * in the debug profile and a `p < first+size` containment rewrite misses
 * the record whose first+size wraps to 0.
 *
 * Cases (all queries against the REAL ScopeLocal maptable of function
 * GetStr, symbols installed through the public production path
 * Scope::addSymbol with an invalid usepoint — address-tied storage,
 * subsort (0,0)):
 *
 *   wrap_top_full       findOverlap(stack, 0xfffffffffffffff8, 8):
 *                       last = -8 + 8 - 1 wraps in uint8 domain... it does
 *                       NOT wrap: -8+8 = 0 offset 0xfffffffffffffff8 + 8
 *                       = 0x1_0000000000000000 wraps to 0, then -1 wraps
 *                       back to 0xffffffffffffffff. Answer: top8.
 *   wrap_top_last_byte  findOverlap(stack, 0xffffffffffffffff, 1):
 *                       point+size wraps to 0, -1 wraps to
 *                       0xffffffffffffffff. Answer: top8.
 *   wrap_top_from_below findOverlap(stack, 0xfffffffffffffff6, 4):
 *                       [..f6, ..f9] overlaps [..f8, ..ff]. Answer: top8.
 *   neg_size_below      findOverlap(stack, 0x1000, -8):
 *                       modular last = 0xff7 < point → null.
 *   zero_size_top       findOverlap(stack, 0xfffffffffffffff8, 0):
 *                       last = point - 1 < point → null.
 *   low_hit             findOverlap(stack, 0x104, 4): sanity that the
 *                       normal in-range path keeps answering low8.
 *   nonoverlap_above    findOverlap(stack, 0x10, 8): neither record → null.
 */

#include "bfd_arch.hh"
#include "funcdata.hh"
#include "libdecomp.hh"

#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>

namespace {

using namespace ghidra;
using std::ostringstream;
using std::string;

void dumpOverlap(const string &caseName, AddrSpace *spc, uintb offset,
                 int4 size, const SymbolEntry *entry)
{
  ostringstream out;
  out << "case=" << caseName << "|kind=overlap|query=0x" << std::hex
      << offset << ':' << std::dec << size << "|up=inv";
  if (entry == (const SymbolEntry *)0) {
    out << "|result=null";
  }
  else {
    out << "|result=" << entry->getSymbol()->getName()
        << "|first=0x" << std::hex << entry->getFirst()
        << "|last=0x" << std::hex << entry->getLast()
        << "|off=" << std::dec << entry->getOffset()
        << "|sz=" << std::dec << entry->getSize();
  }
  std::cout << out.str() << '\n';
  std::cout.flush();
}

void runCases(Funcdata &fd)
{
  AddrSpace *stack = fd.getArch()->getStackSpace();
  TypeFactory *types = fd.getArch()->types;
  ScopeLocal *lm = fd.getScopeLocal();

  // Address-tied symbols (invalid usepoint → empty uselimit, subsort
  // (0,0), database.cc:1149-1150): top8 = the top-of-stack 8-byte record
  // [0xfffffffffffffff8, 0xffffffffffffffff], low8 = [0x100, 0x107].
  lm->addSymbol("top8", types->getBase(8, TYPE_INT),
                Address(stack, 0xfffffffffffffff8ULL), Address());
  lm->addSymbol("low8", types->getBase(8, TYPE_INT),
                Address(stack, 0x100), Address());

  dumpOverlap("wrap_top_full", stack, 0xfffffffffffffff8ULL, 8,
              lm->findOverlap(Address(stack, 0xfffffffffffffff8ULL), 8));
  dumpOverlap("wrap_top_last_byte", stack, 0xffffffffffffffffULL, 1,
              lm->findOverlap(Address(stack, 0xffffffffffffffffULL), 1));
  dumpOverlap("wrap_top_from_below", stack, 0xfffffffffffffff6ULL, 4,
              lm->findOverlap(Address(stack, 0xfffffffffffffff6ULL), 4));
  dumpOverlap("neg_size_below", stack, 0x1000, -8,
              lm->findOverlap(Address(stack, 0x1000), -8));
  dumpOverlap("zero_size_top", stack, 0xfffffffffffffff8ULL, 0,
              lm->findOverlap(Address(stack, 0xfffffffffffffff8ULL), 0));
  dumpOverlap("low_hit", stack, 0x104, 4,
              lm->findOverlap(Address(stack, 0x104), 4));
  dumpOverlap("nonoverlap_above", stack, 0x10, 8,
              lm->findOverlap(Address(stack, 0x10), 8));
}

void run(const string &specDirectory, const string &binary)
{
  vector<string> specPaths;
  specPaths.push_back(specDirectory);
  startDecompilerLibrary(specPaths);
  {
    BfdArchitecture architecture(binary, "default", &std::cerr);
    DocumentStorage store;
    architecture.init(store);
    architecture.readLoaderSymbols("::");
    Funcdata *fd = architecture.symboltab->getGlobalScope()->queryFunction("GetStr");
    if (fd == (Funcdata *)0)
      throw std::runtime_error("GetStr was not found in the BFD symbol table");
    if (fd->getName() != "GetStr" ||
        fd->getAddress().getOffset() != 0x36d0 || fd->getSize() != 0)
      throw std::runtime_error("GetStr input identity drifted");
    if (architecture.archid != "x86:LE:64:default:gcc")
      throw std::runtime_error("runtime architecture/compiler drifted: " +
                               architecture.archid);
    runCases(*fd);
  }
  shutdownDecompilerLibrary();
}

} // anonymous namespace

int main(int argc, char **argv)
{
  if (argc != 3) {
    std::cerr << "usage: funcdata_scopelocal_wrap_1204 SPEC_ROOT CURL_BINARY\n";
    return 2;
  }
  try {
    run(argv[1], argv[2]);
    return 0;
  }
  catch (const LowlevelError &error) {
    std::cerr << "funcdata_scopelocal_wrap_1204: LowlevelError: "
              << error.explain << '\n';
    return 1;
  }
  catch (const std::exception &error) {
    std::cerr << "funcdata_scopelocal_wrap_1204: " << error.what() << '\n';
    return 1;
  }
}
